.github/workflows/ci.yaml

```diff
@@ -6,64 +8,121 @@ on:
   merge_group:
     types: [checks_requested]
 
-name: Continuous integration
-
 env:
   CARGO_TERM_COLOR: always
 
 jobs:
-  test:
-    name: Test Suite
+  build-test-artifacts:
+    name: Test Artifacts
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v4
-        with:
-          submodules: recursive
-      - uses: dtolnay/rust-toolchain@stable
-      - run: cargo build --all --all-features
-      - run: cargo test --all-features -- --nocapture
+      - name: Checkout PR
+        uses: actions/checkout@v4
 
-  fmt:
-    name: Rustfmt
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v4
-        with:
-          submodules: recursive
-      - uses: dtolnay/rust-toolchain@nightly
+      - name: Use Rust stable
+        uses: dtolnay/rust-toolchain@stable
+
+      - name: Install cargo-llvm-cov
+        uses: taiki-e/install-action@cargo-llvm-cov
+
+      - name: Install nextest
+        uses: taiki-e/install-action@nextest
+
+      - name: Build and archive tests
+        run: |
+          cargo llvm-cov nextest-archive \
+            --all-features \
+            --workspace \
+            --archive-file nextest-archive.tar.zst
+
+      - name: Upload archive to workflow
+        uses: actions/upload-artifact@v3
         with:
-          components: rustfmt
-      - run: cargo fmt -- --config group_imports=StdExternalCrate,imports_granularity=Module --check
+          name: nextest-archive
+          path: nextest-archive.tar.zst
 
-  clippy:
-    name: Clippy
+  run-tests:
+    name: Run Tests
     runs-on: ubuntu-latest
+    needs: build-test-artifacts
     steps:
-      - uses: actions/checkout@v4
-        with:
-          submodules: recursive
-      - uses: dtolnay/rust-toolchain@stable
+      - name: Checkout PR
+        uses: actions/checkout@v4
+
+      - name: Install cargo-llvm-cov
```

Comment:
see above about using a cache - one of the extra benefits is that you can install once and ensure you're using the same crate across all steps. caveat being that error checking the cache exists becomes necessary